### ISSUE TITLE
How to read the `javax.persistence.Column.length()` value during test data generation?

### ISSUE THREAD

**Issue by armandino** [instancio/instancio#474]
### Discussed in https://github.com/orgs/instancio/discussions/467

<div type='discussions-op-text'>

<sup>Originally posted by @nedis-ua February 27, 2023</sup>
Hello,

I would like to generate JPA entities with string test data that are valid according to "`@Column(length = x)`" restrictions.
I.e. something like this:

```
return Instancio.of(entityClass)
    // Try to parse @Column.length() value for test string data:
    .generate(
        fields(field -> field.getType().equals(String.class) && field.isAnnotationPresent(Column.class)),
        gen -> {
            Field field = //TODO getCurrentField instance
            return gen.string().maxLength(field.getAnnotation(Column.class).length());
        })
    .create();
```
Is it possible to set such string generator?</div>

**Comment by armandino** [comment]
@nedis-ua, I added some changes in #479 that I think should solve the issue you're facing. Let me know if you have any feedback on this.

The proposed solution is to implement the `InstancioServiceProvider` interface and register it under `META-INF/services`, e.g.

```java
    class InstancioServiceProviderImpl implements InstancioServiceProvider {

        @Override
        public GeneratorProvider getGeneratorProvider() {
            return (node, generators) -> {
                if (node.getField() != null) {
                    Column column = node.getField().getAnnotation(Column.class);
                    if (column != null) {
                        return generators.string().maxLength(column.length());
                    }
                }
                return null;
            };
        }
    }
```

Then `Instancio.create(entityClass)` will automatically apply the above spec to any field annotated with `@Coiumn`.

**Comment by nedis-ua** [comment]
Hi @armandino 

I verified Your solution and it works. Thank you. 
I am waiting until a new version of `Instancio` lib is released

**Comment by fr3sh** [comment]
I'm also waiting for this feature. 
The same thing will be for example to handle the 
@Size(min = 2, max = 100) 
not from Jakarta but from javax.validation

**Comment by armandino** [comment]
The new version `2.11.0` should be available from Maven central.

@fr3sh `javax` annotations are not supported out-of-the-box, thanks to a PR by @nedis-ua.

@nedis-ua I will close this issue. Appreciate your contributions! :)

**Comment by armandino** [comment]
@Mobe91 thank you for the suggestion and the sample code.

Have you considered resolving the method name from the field name? This is how method reference selectors currently work. For example, when you use `field(Foo::getValue)` it maps the method name `getValue` to field name `value`. You could do this in reverse:

``` java
Field field = node.getField();
PropertyDescriptor pd = new PropertyDescriptor(field.getName(), field.getDeclaringClass());
Method getter = pd.getReadMethod();
Annotation[] annotations = getter.getAnnotations();
// ...
```

In any case, it's an interesting idea to be able to pass user-defined keys to `Settings`. Currently, `Settings` only supports a known set of internally-defined keys. I'll need to think about this some more.

Regarding your last question, all selector interfaces only expose methods that are applicable to the API. Otherwise, IDE auto-completion will start showing suggestions like `field(...).getSettings()`. For this reason, I'm reluctant to do this.

Are you changes available in a public repo? If so, I'd be interested in taking a look.

**Comment by Mobe91** [comment]
> In any case, it's an interesting idea to be able to pass user-defined keys to Settings. Currently, Settings only supports a known set of internally-defined keys. I'll need to think about this some more.

I've already created a custom key in my local setup and it worked 😅

> Have you considered resolving the method name from the field name? This is how method reference selectors currently work. For example, when you use field(Foo::getValue) it maps the method name getValue to field name value. You could do this in reverse:

The problem with this is that it does not consider overriden getters in subtypes which you commonly have in JPA entity models. Consider this example:

```java
public abstract class BaseEntity<IdType> {
  private IdType id;

  public IdType getId() {
    return id;
  }
}

@Entity
public class Car implements BaseEntity<Long> {
  
  @Override
  @Id
  public Long getId() {
    return super.getId();
  }
}
```
Now consider `Instancio.create(Car.class)`.
For the id field in above example, `field.getName()` would evaluate to `id` and `field.getDeclaringClass()` would evaluate to BaseEntity (even though the target type is Car). Hence, the resolved getter will be the one from BaseEntity and any annotations on the method override in the concrete subtype will be missed.

> Regarding your last question, all selector interfaces only expose methods that are applicable to the API. Otherwise, IDE auto-completion will start showing suggestions like field(...).getSettings(). For this reason, I'm reluctant to do this.

All good, I found a way to work around this one.

However, it would still be good to be able to access the settings from within a `InstancioServiceProvider` as mentioned in my [first comment](https://github.com/instancio/instancio/issues/474#issuecomment-1471907743).

> Are your changes available in a public repo? If so, I'd be interested in taking a look.

Not yet but I can make them available. Still finishing up my work. Sneak peak of what I have atm:

```java
Car car = aCar()
            // Pass JPA metamodel, this activates a number of transformers to sanitize the final object graph
            .forPersistence(entityManager.getMetamodel())
            // Use usual Instancio API
            .set(field(Car::getColor), "Blue")
            .set(field(Wheel::getRadius), 120)
            .create();
entityGraphPersister.persist(car);
```
Where `aCar()` returns a facade of `org.instancio.InstancioApi` with an extension for supplying the JPA metamodel. 

I had to make some minor changes to Instancio to make it work. Most importantly

* introduced `org.instancio.internal.selectors.SelectorTargetKind#NODE`
* extended `org.instancio.Node` with a `Class<?> getParentTargetClass();`

**Comment by armandino** [comment]
> I've already created a custom key in my local setup and it worked sweat_smile

Less work for me then! :)

> The problem with this is that it does not consider overriden getters in subtypes which you commonly have in JPA entity models.

I can't think of a use case for processing the `@Id` annotation, but I see your point.

> However, it would still be good to be able to access the settings from within a `InstancioServiceProvider` as mentioned in my [first comment](https://github.com/instancio/instancio/issues/474#issuecomment-1471907743).

The problem with this is that the `Generators` class is also exposed to the `generate()` method. Adding `getContext()` to it would allow usage like `generate(field(...), gen -> gen.getContext())`, which is not valid. For now, I'd suggest using the workaround. I might add something to support this if there is more interest from users.

> Not yet but I can make them available. Still finishing up my work. Sneak peak of what I have atm:

Looks great! It would be handy for those working with JPA entities. Definitely consider publishing your fork :+1:

**Comment by armandino** [comment]
@Mobe91 I made some changes to the SPI in #505 that I think will solve the issue you're facing and doesn't involve changing `Generators`. The idea was to add an `init(ServiceProviderContext context)` to the SPI from which you can get an instance of the `Settings`. It's not as convenient as calling `gen.getContext()`, but I think it's a better alternative overall and doesn't require changing the existing APIs. I'll need a bit more time to test its usage more thoroughly, but in the meantime let me know if you have any feedback on this approach.

**Comment by armandino** [comment]
@Mobe91 the above change is available in the latest version `2.12.0`. Hope it works for your use case.
